### ISSUE TITLE
Equivalence relations for Wedge, Smash, Can

### ISSUE THREAD

**Issue by subttle** [emilypi/smash#23]
Hi again,
I've written equivalence relations (using `Equivalence` from `Data.Functor.Contravariant`) for `Wedge`, `Smash`, and `Can`, (and also `These` but I'll submit that elsewhere) and have found them useful in my own code, so I thought it would be worth mentioning in case it seems like something you would like to make available in your package:

```Haskell
import Data.Functor.Contravariant (Equivalence (..))

byWedge ∷ Equivalence (Wedge a b)
byWedge = Equivalence (≡)
  where
    (≡) ∷ Wedge a b → Wedge a b → Bool
    (≡) (Nowhere  ) (Nowhere  ) = True
    (≡) (Here    _) (Here    _) = True
    (≡) (There   _) (There   _) = True
    (≡) _           _           = False

bySmash ∷ Equivalence (Smash a b)
bySmash = Equivalence (≡)
  where
    (≡) ∷ Smash a b → Smash a b → Bool
    (≡) (Nada     ) (Nada     ) = True
    (≡) (Smash _ _) (Smash _ _) = True
    (≡) _           _           = False

byCan ∷ Equivalence (Can a b)
byCan = Equivalence (≡)
  where
    (≡) ∷ Can a b → Can a b → Bool
    (≡) (Non    ) (Non    ) = True
    (≡) (One   _) (One   _) = True
    (≡) (Eno   _) (Eno   _) = True
    (≡) (Two _ _) (Two _ _) = True
    (≡) _           _       = False

 ```
 Here is an arbitrary example use:
 
```Haskell
import Data.Functor.Contravariant (contramap)

graphComponents ∷ ∀ a . (Num a, Eq a) ⇒ Equivalence (a, a)
graphComponents = contramap getComponents byCan
  where
    getComponents ∷ (a, a) → Can a a
    getComponents (0, 0) = Non     -- origin
    getComponents (x, 0) = One x   -- x-axis
    getComponents (0, y) = Eno   y -- y-axis
    getComponents (x, y) = Two x y -- quadrant
```
Again, I won't be offended if you don't like the idea, just sharing in case you do. If you do then let me know if you'd like a pull request (and of course I'm open to suggestions for renaming). I've only written test coverage for finite cases, so I may have to think about how to write tests without a `Finite` constraint, or instead I could do a few quick proofs in Agda so I could avoid dealing with any `Finite` constraints.

Hope you have a great day, thanks again for all the hard work you put into this package!

**Comment by emilypi** [comment]
I'd welcome a PR for this.

**Comment by subttle** [comment]
Working on this now and giving a quick update, per our earlier discussion I've replaced `byWedge`, `bySmash`, and `byCan` with `eqWedge`, `eqSmash`, and `eqCan` respectively (thanks for the great suggestion). For placement I've chosen to include each under the section labeled with the "Partitioning" comment, due to the [special](https://proofwiki.org/wiki/Fundamental_Theorem_on_Equivalence_Relations) [correspondence](https://proofwiki.org/wiki/Relation_Induced_by_Partition_is_Equivalence) of partitioning and equivalence relations.

Also, I should note about the dependency for `contravariant`, according to the docs it has only been in base "Since: base-4.12.0.0". Because GHC-8.4.4 (which depends on `base-4.11`) is currently supported, I'll have to bring in the dependency for the `contravariant` package, but if you decide to drop support for GHC-8.4.4 for whatever reason in the future then you could remove said dependency and just bump the `base` to `>=4.12`.

**Comment by emilypi** [comment]
I'm fine with bumping to 4.12

**Comment by subttle** [comment]
Okay, cool. Would you prefer I altogether remove:
```
          - ghc: 8.4.4
            allow-failure: false
```
from the `haskell-ci.yml` completely then (and then update each `*.cabal` file for the rest of the sub-directories to remove `tested-with:` GHC-8.4.4 or would you prefer I just set `allow-failure: true` and leave the rest of the `*.cabal` files as-is?

**Comment by emilypi** [comment]
do you have [haskell-ci](https://github.com/haskell-CI/haskell-ci)? If you install it from HEAD, then all you need to do is remove the `tested-with` entry from each subproject and run `haskell-ci regenerate` and it'll do the CI bumps for you. Then, you just up the base bounds on each project and you're good to go

**Comment by emilypi** [comment]
Closed via #28

**Comment by emilypi** [comment]
Thanks @subttle !
